File: alchemist/alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/conditions/AbstractCondition.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

/**
 * 
 */
package it.unibo.alchemist.model.implementations.conditions;

import java.util.Objects;

import it.unibo.alchemist.model.interfaces.Dependency;
import org.danilopianini.util.LinkedListSet;
import org.danilopianini.util.ListSet;

import it.unibo.alchemist.model.interfaces.Condition;
import it.unibo.alchemist.model.interfaces.Node;
import it.unibo.alchemist.model.interfaces.Reaction;


/**
 *
 * @param <T>
 */
public abstract class AbstractCondition<T> implements Condition<T> {

    private static final long serialVersionUID = -1610947908159507754L;
    private final ListSet<Dependency> influencing = new LinkedListSet<>();
    private final Node<T> n;

    /**
     * @param node the node this Condition belongs to
     */
    public AbstractCondition(final Node<T> node) {
        this.n = Objects.requireNonNull(node);
    }

    /**
     * {@inheritDoc}
     * 
     * How to override: if you intend your condition to be potentially changed by
     * any change in the context, return null.
     */
    @Override
    public final ListSet<? extends Dependency> getInboundDependencies() {
        return influencing;
    }

    /**
     * {@inheritDoc}
     * 
     * Override if your {@link Condition} can return a more specific type of node.
     * The typical way is to cast the call to super.getNode().
     */
    @Override
    public Node<T> getNode() {
        return n;
    }

    /**
     * @param m the molecule to add
     */
    protected final void declareDependencyOn(final Dependency m) {
        influencing.add(m);
    }

    /**
     * {@inheritDoc}
     * 
     * How to override: create a new action of your concrete subtype.
     */
    @Override
    public Condition<T> cloneCondition(final Node<T> n, final Reaction<T> r) {
        throw new UnsupportedOperationException(getClass().getSimpleName() + " has no support for cloning.");
    }

    /**
     *
     * @return the simple class name
     */
    @Override
    public String toString() {
        return getClass().getSimpleName();
    }
}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/conditions/AbstractCondition.kt
/*
 * Copyright (C) 2010-2026, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */

package it.unibo.alchemist.model.conditions

import it.unibo.alchemist.model.Condition
import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.NodeReaction
import it.unibo.alchemist.model.observation.MutableObservable
import it.unibo.alchemist.model.observation.Observable

/** Base implementation of a reactive [Condition]. */
open class AbstractCondition<T>(private val node: Node<T>) : Condition<T> {
    private var validity: Observable<Boolean> = MutableObservable.observe(true)

    override fun getNode(): Node<T> = node

    override fun isValid(): Observable<Boolean> = validity

    override fun dispose() {
        validity.dispose()
    }

    override fun cloneCondition(newNode: Node<T>, newReaction: NodeReaction<T>): Condition<T> =
        throw UnsupportedOperationException("${javaClass.simpleName} has no support for cloning.")

    /**
     * Installs the observable backing [isValid].
     *
     * The condition owns the derived view created here, while [newValidity] remains owned by its model source.
     */
    protected fun setValidity(newValidity: Observable<Boolean>) {
        validity.dispose()
        validity = newValidity.map { it }
    }

    override fun toString(): String = javaClass.simpleName
}
